src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs

```diff
@@ -5163,5 +5163,33 @@ public void CompileAndVerifyModuleIncludesAllModules()
                 Assert.Equal("refMod.netmodule", a.ContainingModule.Name);
             });
         }
+
+        [Fact]
+        [WorkItem(37779, "https://github.com/dotnet/roslyn/issues/37779")]
+        public void WarnAsErrorDoesNotEmit()
+        {
+            string source = @"
+class X
+{
+    int _f;
+}";
+            var options = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary, generalDiagnosticOption: ReportDiagnostic.Error);
+            var compilation = CreateCompilation(source, options: options);
+
+            EmitResult emitResult;
+            using (var output = new MemoryStream())
+            {
+                emitResult = compilation.Emit(output, pdbStream: null, xmlDocumentationStream: null, win32Resources: null);
```

Comment:
We also need to verify that the PDB and XML docs are not generated.

Reply:
Updated the test to verify pdb is not generated. XML documentation generation happens [prior](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L1011) to the compiler [force completing](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L1033) the trees, which happens prior to [application of diagnostic suppressors](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L1046). We either need to change this ordering OR special case bailing out of XML documentation generation in presence of warnings escalated to errors + no diagnostic suppressors. I think either of these would add unnecessary complexity to this PR and should be done as a separate change. I can file a separate tracking issue if that seems reasonable.

Reply:
No, I don't think leaving this unfixed is OK.

Reply:
I think the previous behavior was that CompileMethods would have failed, which would have stopped XML generation. If this is broken by diagnostic suppressors, that's not acceptable.

Reply:
@agocke - What is your proposed fix?

Reply:
At the moment I don't have one. If we can't think of one, we should revert the diagnostic suppressor changes.

Reply:
> If we can't think of one, we should revert the diagnostic suppressor changes.

I did propose an option which would cover majority of cases: special case bailing out of XML documentation generation in presence of warnings escalated to errors + no diagnostic suppressors.

Reply:
The compiler can't write doc files if the build fails. Fixing most of the cases isn't good enough.

Reply:
I believe the options are:

1. Revert the DiagnosticSupppressor API completely
2. Make DiagnosticSuppressor API specific to analyzer diagnostics only (compiler warnings cannot be programmatically suppressed)
3. Tweak the current PR so that XML doc file will not be generated for build failures, except in presence of warnings escalated to errors if there is at least one diagnostic suppressor.
4. Keep the PR in current state where XML doc will still get generated if there are no compiler errors, but only compiler warnings escalated to errors.

Tagging @jinujoseph @tmat @sharwell for their views. @agocke believes 3 or 4 is not acceptable.

Reply:
I agree with @agocke that 3 and 4 are not acceptable. When compilation fails no output should occur. That's not an invariant of the compiler we can change for this feature.

Reply:
If there is no plan, or understood design, that will restore the compiler invariants then I think we need to consider option 1. This feature is still fresh and pulling it should impact a relatively small number of people. Also if it's not ready we should remove it before we ship .NET Core 3.0. We can work through the design separately, find the right one and put it back. 

If this was an edge case or if the change necessary to restore the invariants were understood I'd be more content to treat this as a bug and fix in the next release. Reading through the thread though it seems there is not an understanding of how this would be fixed just an expectation that it will be complex.

Reply:
Investigated further and https://github.com/dotnet/roslyn/pull/37948/commits/2c33e99763aa870ed56935a31ccde1e87cff3522 should give the desired behavior and has the requested unit tests.

Reply:
@agocke @jaredpar My [commit](https://github.com/dotnet/roslyn/commit/2c33e99763aa870ed56935a31ccde1e87cff3522) to establish your requested invariant leads to unit test failures in existing compiler unit tests which are validating xml documentation in presence of compiler errors:

1. [WRN_XMLDocInsideMethod](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic.Symbol.UnitTests/DocumentationComments/DocCommentTests.vb,4191)
2. [WRN_XMLDocInsideMethod_NoError](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic.Symbol.UnitTests/DocumentationComments/DocCommentTests.vb,4262)
3. [ExtendedCref_BinaryOperator](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic.Symbol.UnitTests/DocumentationComments/DocCommentTests.vb,10726):  Failing scenario is specifically [this](http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic.Symbol.UnitTests/DocumentationComments/DocCommentTests.vb,10798) one where test expects xml doc output for input code with errors.

These unit tests haven't changed for quite a while, so it seems the invariant has been broken for some time, long before the diagnostic suppressors change. Let me know if you want me to fix these unit tests in this PR or revert my last commit and file a separate tracking issue for compiler team to restore the invariant.

Reply:
Looks like Those methods are calling `Emit` directly. Sounds like we need to enumerate the expected behavior for all of these circumstances for both VB and C# and decide exactly how we want to treat each case (as a bug or as expected behavior).

Reply:
Yes, looking the code, `Emit` does one thing, `CommonCompiler` does another. I see some VB tests that verify the Emit behavior, but not C# tests. Did you find any?

Reply:
Nope, only VB compiler tests. I can tweak my change to just bail out from the doc file generation in Common compiler, but let the Emit API still generate the documentation in the provided xmlDocStream. Alternatively, we can avoid any xml doc generation in presence of errors for both scenarios and just fix the unit tests.

Reply:
Since this is an ask mode change, let's keep the current behavior, add a comment, and file a bug.

Reply:
Also, it looks like there's a failure in NoWarnAndWarnAsError_WarningDiagnostic. Have you figured out what's going on there?

Reply:
And we should add tests to verify all of this behavior.

Reply:
> Since this is an ask mode change, let's keep the current behavior, add a comment, and file a bug.

Just to confirm - we will ensure that common compiler code path does not generate xml doc file, but Emit API invocation generates xml documentation so there are no unit test changes required + file a compiler issue to revisit this behavior. Is that correct understanding?

> Also, it looks like there's a failure in NoWarnAndWarnAsError_WarningDiagnostic. Have you figured out what's going on there?

Yes, my last commit bails out completely prior to xml doc generation in presence of errors. This leads to us skipping reporting any analyzer diagnostics. I was planning to restore the original behavior of reporting analyzer diagnostics even in presence of compiler errors by just guarding xml doc generation but not returning. That should fix those unit tests.

Reply:
> Is that correct understanding?

Yup sounds right to me.

Reply:
Note: it still doesn't seem to generate a PE file after Emit (which is weird) but also seems like existing behavior.

Reply:
Filed https://github.com/dotnet/roslyn/issues/37996.